src/ClientContext.ts

```diff
@@ -91,67 +81,38 @@ export class ClientContext {
     // Query operations will use ReadEndpoint even though it uses
     // GET(for queryFeed) and POST(for regular query operations)
 
-    const request: any = {
-      // TODO: any request
+    const request: RequestContext = {
+      globalEndpointManager: this.globalEndpointManager,
+      requestAgent: this.cosmosClientOptions.agent,
+      connectionPolicy: this.connectionPolicy,
+      method: HTTPMethod.get,
       path,
       operationType: OperationType.Query,
       client: this,
-      endpointOverride: null
+      partitionKeyRangeId,
+      resourceId,
+      resourceType,
+      options,
+      body: query
     };
 
-    const endpoint = await this.globalEndpointManager.resolveServiceEndpoint(request);
+    if (query !== undefined) {
+      request.method = HTTPMethod.post;
+    }
 
-    const initialHeaders = { ...this.cosmosClientOptions.defaultHeaders, ...(options && options.initialHeaders) };
-    if (query === undefined) {
-      const reqHeaders = await getHeaders(
-        this.cosmosClientOptions.auth,
-        initialHeaders,
-        HTTPMethod.get,
-        path,
-        id,
-        type,
-        options,
-        partitionKeyRangeId,
-        this.cosmosClientOptions.connectionPolicy.useMultipleWriteLocations
-      );
-      this.applySessionToken(path, reqHeaders);
-
-      const response = await this.requestHandler.get(endpoint, request, reqHeaders);
-      this.captureSessionToken(undefined, path, OperationType.Query, response.headers);
-      return this.processQueryFeedResponse(response, !!query, resultFn);
-    } else {
-      initialHeaders[Constants.HttpHeaders.IsQuery] = "true";
-      switch (this.cosmosClientOptions.queryCompatibilityMode) {
-        case QueryCompatibilityMode.SqlQuery:
-          initialHeaders[Constants.HttpHeaders.ContentType] = Constants.MediaTypes.SQL;
-          break;
-        case QueryCompatibilityMode.Query:
-        case QueryCompatibilityMode.Default:
-        default:
-          if (typeof query === "string") {
-            query = { query }; // Converts query text to query object.
-          }
-          initialHeaders[Constants.HttpHeaders.ContentType] = Constants.MediaTypes.QueryJson;
-          break;
+    await this.globalEndpointManager.setServiceEndpoint(request);
```

Comment:
should we just move the setServiceEndpoint (or it's equivalent) call into executeRequest? We're passing the globalEndpointManager down there already

Reply:
Yes, we probably can. I'll address that when I do proper pipelines